[GVN] Load-store forwaring of scalable store to fixed load.

iamlouk opened the pull request:
When storing a scalable vector and loading a fixed-size vector, where the
scalable vector is known to be larger based on vscale_range, perform
store-to-load forwarding through temporary @llvm.vector.extract calls.
InstCombine then folds the insert/extract pair away.

The usecase is shown in this [godbolt
link](https://godbolt.org/z/KT3sMrMbd), which shows that clang generates
IR that matches this pattern when the "arm_sve_vector_bits" attribute is
used:

```c
typedef svfloat32_t svfloat32_fixed_t
__attribute__((arm_sve_vector_bits(512)));

struct svfloat32_wrapped_t {
  svfloat32_fixed_t v;
};

static inline svfloat32_wrapped_t
add(svfloat32_wrapped_t a, svfloat32_wrapped_t b) {
  return {svadd_f32_x(svptrue_b32(), a.v, b.v)};
}

svfloat32_wrapped_t
foo(svfloat32_wrapped_t a, svfloat32_wrapped_t b) {
  // The IR pattern this patch matches is generated for this return:
  return add(a, b);
}
```

iamlouk commented:
@davemgreen If you would not mind, could you have a look at this? Thanks in advance!

nikic reviewed (commented):
Looks reasonable from a cursory look.

I think this could be straightforwardly generalized to the case where store >= load rather than store == load (in which case we don't need to know vscale exactly, just the minimum). In that case it's still correct to extract the prefix.

nikic commented on llvm/test/Transforms/GVN/vscale.ll:
It would be good to add a negative test where the load has an extra constant offset, so make sure we don't forward in that case.

davemgreen replied:
Different types might be useful too. For example a scalable_store_to_fixed_load with `<vscale x 4 x float>` input but `<vscale x 4 x i32>` output for example.

iamlouk replied:
I added one just now (in the existing test file).

iamlouk commented on llvm/lib/Transforms/Utils/VNCoercion.cpp:
```diff
@@ -79,14 +100,23 @@ bool canCoerceMustAliasedValueToLoad(Value *StoredVal, Type *LoadTy,
                                       const DataLayout &DL) {
-  assert(canCoerceMustAliasedValueToLoad(StoredVal, LoadedTy, DL) &&
+  assert(canCoerceMustAliasedValueToLoad(
+             StoredVal, LoadedTy, Helper.GetInsertBlock()->getParent()) &&
```
I changed that here, but not in functions like `coerceAvaialableValueToLoadType` because as long as the offset is restricted to be zero when scalable vectors are involved, all of those only need the DataLayout. I can of course change the interface everywhere for consistency, even in functions that only need the DL.

iamlouk commented on llvm/test/Transforms/GVN/vscale.ll:
```diff
@@ -641,3 +641,63 @@ entry:
+  ret <vscale x 4 x float> %cast.scalable
+}
+
+define <vscale x 4 x float> @scalable_store_to_fixed_load_unknon_vscale(<vscale x 4 x float> %.coerce) {
```
Sorry, thanks, fixed!

iamlouk commented:
> I think this could be straightforwardly generalized to the case where store >= load rather than store == load (in which case we don't need to know vscale exactly, just the minimum). In that case it's still correct to extract the prefix.

I just updated the MR and added that functionality.

davemgreen reviewed (commented):
> @davemgreen If you would not mind, could you have a look at this? Thanks in advance!

I just happened to be the last person who touched this, in #123984. One thing that I thought was probably useful was to add NewGVN test coverage, to make sure it doesn't do anything unexpected.

iamlouk committed:
[GVN] Tests for load-store forwaring of scalable store to fixed load

nikic commented on llvm/lib/Transforms/Utils/VNCoercion.cpp:
```diff
@@ -13,32 +13,52 @@ static bool isFirstClassAggregateOrScalableType(Type *Ty) {
+
+    // If the VScale is known at compile-time, use that information to
+    // allow for wider loads.
+    std::optional<unsigned> VScale = getKnownVScale(F);
```
This could be generalized to only look at getVScaleRangeMin to determine the minimum store size (without knowing the exact store size), as we only need StoreSize >= LoadSize. (Don't know if we expect that pattern to occur.)

iamlouk replied:
Thanks, I did add this generalization, and therefore removed the `getKnownVScale(...)` helper and also renamed the calculated size to `MinStoreSize`.

iamlouk replied:
I just added it.

iamlouk committed:
[GVN] Load-store forwaring of scalable store to fixed load.

When storing a scalable vector and the vscale is a compile-time known
constant, store-to-load forwarding through temporary
@llvm.vector.extract calls, even if the loaded vector is fixed-sized
instead of scalable. InstCombine then folds the insert/extract pair
away.

The usecase is shown in this [godbold
link](https://godbolt.org/z/KT3sMrMbd), which shows that clang generates
IR that matches this pattern when the "arm_sve_vector_bits" attribute is
used:

```c
typedef svfloat32_t svfloat32_fixed_t
__attribute__((arm_sve_vector_bits(512)));

struct svfloat32_wrapped_t {
  svfloat32_fixed_t v;
};

static inline svfloat32_wrapped_t
add(svfloat32_wrapped_t a, svfloat32_wrapped_t b) {
  return {svadd_f32_x(svptrue_b32(), a.v, b.v)};
}

svfloat32_wrapped_t
foo(svfloat32_wrapped_t a, svfloat32_wrapped_t b) {
  // The IR pattern this patch matches is generated for this return:
  return add(a, b);
}
```

nikic commented on llvm/test/Transforms/GVN/vscale.ll:
```diff
@@ -641,3 +641,117 @@ entry:
   ret { <vscale x 16 x i8>, <vscale x 16 x i8>, <vscale x 16 x i8>, <vscale x 16 x i8> } %15
 }
+
+define <vscale x 4 x float> @scalable_store_to_fixed_load(<vscale x 4 x float> %.coerce) #1 {
```
Nit: I think these tests are easier to read if you inline the vscale range.
```suggestion
define <vscale x 4 x float> @scalable_store_to_fixed_load(<vscale x 4 x float> %.coerce) vscale_range(4,4) {
```

iamlouk replied:
Thanks, I just did that.

nikic commented on llvm/test/Transforms/GVN/vscale.ll:
```diff
@@ -641,3 +641,117 @@ entry:
+
+declare <vscale x 4 x float> @llvm.vector.insert.nxv4f32.v16f32(<vscale x 4 x float>, <16 x float>, i64 immarg)
+declare <vscale x 4 x i32> @llvm.vector.insert.nxv4i32.v16i32(<vscale x 4 x i32>, <16 x i32>, i64 immarg)
+declare <vscale x 4 x float> @llvm.vector.insert.nxv4f32.v32f32(<vscale x 4 x float>, <32 x float>, i64 immarg)
```
nit: You can drop these declarations, they're not required.

iamlouk replied:
Thanks, I just did that.

iamlouk commented:
> One thing that I thought was probably useful was to add NewGVN test coverage, to make sure it doesn't do anything unexpected.

NewGVN does not seam to use the `canCoerceMustAliasedValueToLoad` mechanism yet, I added all the tests for the normal GVN pass there as well, but the added functionality of this MR does not apply there.

nikic reviewed (approved):
LGTM
